cmd/integration/harness.go

```diff
@@ -0,0 +1,194 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/spacemeshos/go-spacemesh/api/pb"
+
+	"google.golang.org/grpc"
+)
+
+const execPathLabel = "executable-path"
+
+// Contains tells whether a contains x.
+// if it does it returns it's index otherwise -1
+func Contains(a []string, x string) int {
+	for ind, n := range a {
+		if strings.Contains(n, x) {
+			return ind
+		}
+	}
+
+	return -1
+}
+
+// Harness fully encapsulates an active node server process
+// along with client connection and full api, created for rpc integration
+// tests and may be used for any other purpose.
+type Harness struct {
+	server *server
+	conn   *grpc.ClientConn
+	pb.SpacemeshServiceClient
+}
+
+func NewHarnessDefaultServerConfig(args []string) (*Harness, error) {
+	// same as in suite's yaml file
+	// find executable path label in args
+	execPathInd := Contains(args, execPathLabel)
+	if execPathInd == -1 {
+		return nil, fmt.Errorf("could not find executable path in arguments")
+	}
+	// next will be exec path value
+	execPath := args[execPathInd+1]
+	// remove executable path label and value
+	args = append(args[:execPathInd], args[execPathInd+2:]...)
+	// set servers' configuration
+	cfg, errCfg := DefaultConfig(execPath)
+	if errCfg != nil {
+		return nil, fmt.Errorf("failed to build mock node: %v", errCfg)
+	}
+	return NewHarness(cfg, args)
+}
+
+// NewHarness creates and initializes a new instance of Harness.
+func NewHarness(cfg *ServerConfig, args []string) (*Harness, error) {
+	fmt.Println("Starting harness")
+	server, err := newServer(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// kill node process in case it is already up
+	if isListening(cfg.rpcListen) {
+		err = killProcess(cfg.rpcListen)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Spawn a new mockNode server process.
+	fmt.Println("harness passing the following arguments:\n", args)
+	fmt.Println("Full node server start listening on:", server.cfg.rpcListen+"\n")
+	if err := server.start(args); err != nil {
+		fmt.Println("Full node ERROR listening on:", server.cfg.rpcListen+"\n")
+		return nil, err
+	}
+
+	//// Verify the client connectivity.
+	//// If failed, shutdown the server.
+	//conn, err := connectClient(cfg.rpcListen)
+	//if err != nil {
+	//	_ = server.shutdown()
+	//	return nil, err
+	//}
+
+	h := &Harness{
+		server: server,
+		//conn:                   conn,
+		//SpacemeshServiceClient: pb.NewSpacemeshServiceClient(conn),
+	}
+
+	return h, nil
+}
+
+// TearDown stops the harness running instance.
+// The created process is killed
+func (h *Harness) TearDown() error {
```

Comment:
Who calls this function?

Reply:
we can if we'd like in the future and decide what happens when we kill the harness.
again I can remove it, not using it at the moment

Reply:
fixed in 1b6eba9f2b0bfa81d747b0bc801c3f2c1fd0a2d0